File: Python/binaryTree/114.flatten-binary-tree-to-linked-list.py
#
# @lc app=leetcode id=114 lang=python3
#
# [114] Flatten Binary Tree to Linked List
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if (root is None): return 

        rightLinkList = self.flatten(root.right) 
        leftLinkList = self.flatten(root.left)  

        if (leftLinkList):
            # 连接root和左链表
            root.right  = leftLinkList 
            leftLinkList.left = None
            # 连接左右链表
            while(leftLinkList.right):
                leftLinkList = leftLinkList.right
            leftLinkList.right = rightLinkList
        else: 
            root.right = rightLinkList    
        root.left = None 
        return root
        
# @lc code=end


    # 如果左边的链表存在把root连接到左边链表的表头
    # 接着遍历到左边链表的结尾，把左边链表的结尾连接到右边链表的表头

File: LeetCode_Go/BinaryTree/114.flatten-binary-tree-to-linked-list.go
/*
 * @lc app=leetcode id=114 lang=golang
 *
 * [114] Flatten Binary Tree to Linked List
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Flatten a binary tree into a linked list using its pre-order traversal, where the linked list uses the TreeNode class with Right child pointing to the next node and Left child always null.

// DFS, post-order traversal. Each node receives a flattened left tree and a flattened right tree, with pointers to the last node (tails) in the linked lists. Then append the left linked list to the Right of root, and append the right linked list to the last node in the left linked list. Return the correct tails from leftTail, rightTail or root.

// O(n) time, O(n) space.

func flatten(root *TreeNode) {
	_ = flattenAndReturnTail(root)
	return
}

func flattenAndReturnTail(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	leftTail := flattenAndReturnTail(root.Left)
	rightTail := flattenAndReturnTail(root.Right)

	left := root.Left
	right := root.Right

	root.Left = nil

	// if there is a left tree, append the right flattened tree to the leftTail, and reconnect the root.Right to the left tree
	if left != nil {
		leftTail.Right = right
		root.Right = left
	}

	// if there is a right tree, return the right tree's tail as rightTail, otherwise return leftTail
	if rightTail != nil {
		return rightTail
	} else if leftTail != nil {
		return leftTail
	} else {
		return root
	}
}

// @lc code=end

